### DISCUSSION TITLE
Idea/Suggestion if possible: Restore Volume Control options somewhere

### DISCUSSION THREAD

**Discussion by Gaurav-Original-ClassicShellTester** [valinet/ExplorerPatcher#188] [Ideas]
In Windows 7, SndVol.exe added a -P switch which opened Volume Control options so we could change volume of multiple audio outputs from a single flyout.

![image](https://user-images.githubusercontent.com/46422834/138551359-e96f0277-c39b-4f61-bbcf-91e2985a88c9.png)

This was also in the right click menu of the Volume icon on the Taskbar but was removed in Windows 10 for no reason. Is there some way to restore it?

The -P switch itself still works in Windows 10 and 11. Just the menu item was silently removed by Microsoft.

**Comment by Anixx** [comment]
LOL I did not know this function existed.

**Reply by valinet** [reply]
Me too. It was hidden away so that’s why I came to prefer the Windows 10 volume switcher which kind of added that via the drop down. @Gaurav-Original-ClassicShellTester Idk, devise a plan on how this should be implemented.

**Reply by Gaurav-Original-ClassicShellTester** [reply]
IDK either. I see, so adding a menu item is tricky indeed. Can that useless "Troubleshoot sound problems" menu item be converted into the more useful Volume Control options? That troubleshooter can be started from Windows Settings app, Control Panel and Start menu search. But IDK if the string can be renamed and what happens about localization.

**Reply by valinet** [reply]
Maybe, yeah, it's doable. There's the issue of locating a proper translation for that in the system files as well, so it appears okay on all systems... I have a tool which can dump localized strings from DLLs [here](https://github.com/valinet/ExplorerPatcher/discussions/183), maybe you can use that to find something suitable as well. But yeah, we could try replacing that link, sure. And provide me with the full command to get to that pop-up, please.

**Reply by valinet** [reply]
Or we could replace the single line flyout, or provide an option to displaying that multi column one or the single column one... idk.

**Reply by Gaurav-Original-ClassicShellTester** [reply]
The full command to open it is simply:
%windir%\system32\sndvol.exe -p

There you select the audio out endpoints for which you wish to adjust volume and click OK. Then they always appear when you single click the volume icon in the tray.

But the list itself is dynamic. The audio endpoints change and appear/disappear depending on whether they are connected/disabled/disconnected. e.g. if some Bluetooth audio device is connected, it appears. Laptop speakers always appear but line out/speaker out only if the 3.5mm jack is connected. Many virtual audio endpoints also appear if apps like Virtual Audio Cable or DroidCam are installed.

**Reply by valinet** [reply]
Ok, I understand, so it is another config panel. Yeah, maybe showing another option only when you enabled the classic volume flyout, as it doesn't make sense for the others. So yeah, look for a suitable translation string in the system files and give me the ID, something that doesn't duplicate what's already in the menu either. The volume flyout applet is `SndVolSSO.dll`. String 2033 from in there may be a candidate if nothing better surfaces.

**Reply by Gaurav-Original-ClassicShellTester** [reply]
Yes, in 7 it was here:
![image](https://user-images.githubusercontent.com/46422834/138558721-92797616-d927-48bb-ba4a-8be20a4d4bdd.png)

But even after running dumpstringtable on sndvolsso.dll as well as sndvol.exe and mmsys.cpl, I couldn't locate that string "Volume Control Options". The only string with the word "options" was "Policy options" in mmsys. I ran dumpstringtable on both Windows 7's files as well as Windows 10!

Anyway, another way I know to find translated strings is Microsoft's Language Portal. I found "Volume Control Options" text there: :D

https://www.microsoft.com/en-us/language/Search?&searchTerm=Volume%20Control%20Options&langID=623&Source=true&productid=0

**Reply by valinet** [reply]
Yeah, it doesn't mean a string lives in the same DLL using it. The portal is nice and very Appley (never knew about it) and actually good for confirming something is indeed in there, but we need the actual location where that string is located, otherwise I can't use it. 

Idk, maybe that tool (DumpStringTable) is wrong or incomplete or doesn't cover 100% of the cases, I don't know 100% for sure how everything works, but we still need somehow to find where that string is exactly located. Try looking for it in some other file, as I said, DLLs can pull data from other DLLs and so on, it's a giant maze basically. If the portal says it's in there, it probably is in some DLL.

**Reply by Anixx** [reply]
For me the command `DumpStringTable C:\windows\system32\sndvol.exe >dst.txt` returns this:

```
id:500: ?????????
id:501: ????????? ???? "%1"
id:502: %1: ?????????
id:503: %1: ????????? ????
id:504: ??? ?????? ?? ?????
id:505: ???????? ?????????? ????????? ??? "%s"
id:506: ?????? ????????? - %s
id:508: %1: ???????? ????
id:509: ????????????? ?? %1
```

**Reply by valinet** [reply]
`cmd` has some issue displaying characters other than ASCII. You have your Windows set in Russian, that's why. Dump to file (`> test.txt`) maybe?

**Reply by valinet** [reply]
Or simply use Resource Hacker.

**Reply by Anixx** [reply]
I did so, dumped it. This is a copypaste from the dumped file.

**Reply by Anixx** [reply]
Resource Hacker does not help.

**Reply by valinet** [reply]
It does, it can show string tables as well.

**Reply by Anixx** [reply]
![изображение](https://user-images.githubusercontent.com/1271736/138561501-623a0355-ef75-4e00-93c5-79d7162ca08d.png)

**Reply by valinet** [reply]
Try this:

[DumpStringTable.zip](https://github.com/valinet/ExplorerPatcher/files/7403197/DumpStringTable.zip)

Use like so:

```
DumpStringTable C:\windows\system32\sndvol.exe dst.txt
```

2nd argument should be the output file name.

**Reply by valinet** [reply]
@Anixx 

...

It's not in the EXE, it's in the MUI file as @Gaurav-Original-ClassicShellTester hinted earlier. The whole idea with localization is to separate language files from the binaries. `DumpStringTable` loads the said file in process as datafile and that automatically has Windows resolve the localizable stuff from whatever file it determines should be resolved from, probably the user language, Russian in your case. That's why it works directly on the EXE. It's more convenient, indeed. Resource Hacker on the other hand just shows you what's in the actual file, which doesn't contain any strings.

**Reply by Anixx** [reply]
So, where is the English mui for this file? My system32 only has Russian one.

As to the tool, I get VCRUNTIME140_1.dll is missing. Not sure what to do, I have VC 2015 redistributable already installed.

**Reply by valinet** [reply]
Use this, it works now:

[DumpStringTable.zip](https://github.com/valinet/ExplorerPatcher/files/7403211/DumpStringTable.zip)

**Reply by valinet** [reply]
@Anixx Does it work now?

**Reply by Anixx** [reply]
No. Still those question marks in the output file.

**Reply by Anixx** [reply]
Do not bother with this please. This is not an important issue. :-)

**Reply by valinet** [reply]
Try this as well:

[DumpStringTable.zip](https://github.com/valinet/ExplorerPatcher/files/7403227/DumpStringTable.zip)

**Reply by Anixx** [reply]
Yes, this one works! Both as dumping into a file and as a console output! 
There is no the above-mentioned string in the file though, but I am glad you have already found it.
![изображение](https://user-images.githubusercontent.com/1271736/138562462-a38483fe-54eb-41d9-bc73-60f3687e0d3b.png)

**Reply by valinet** [reply]
Fuck Microsoft and their convoluted Unicode support. Solution is from:

https://stackoverflow.com/questions/2492077/output-unicode-strings-in-windows-console-app

Using the wide string everywhere is apparently not enough, you also have to set make the runtime library functions understand those as well.

**Reply by valinet** [reply]
So yeah, use the ">" operator from now on, no need for the app to support dumping to file directly.

**Reply by valinet** [reply]
One last try, so I know if I can publish this:

[DumpStringTable.zip](https://github.com/valinet/ExplorerPatcher/files/7403241/DumpStringTable.zip)

To output to file, use ">" as earlier.

**Reply by Gaurav-Original-ClassicShellTester** [reply]
Yes now the question marks have vanished and it dumps it to the .txt file with >

https://i.imgur.com/nVggngF.png

**Reply by valinet** [reply]
What question marks in English anyway, that's full ASCII, shouldn't have had a problem then... anyway, glad that it works now.

**Reply by Anixx** [reply]
Yes, it works.

**Reply by Gaurav-Original-ClassicShellTester** [reply]
But with earlier version, I got questions marks too for a lot of strings (which are now fixed) :)

**Reply by Anixx** [reply]
![изображение](https://user-images.githubusercontent.com/1271736/138562830-d5dae4e8-0cb7-414e-a74b-8355c56c06a9.png)

**Reply by valinet** [reply]
> 
> 
> But with earlier version, I got questions marks too for a lot of strings (which are now fixed) :)

Yeah, seems I did too, whatever, now the standard library is properly set as well, so it works. Problem solved.

**Comment by Gaurav-Original-ClassicShellTester** [comment]
OK found it in Windows 7's sndvolsso.dll.mui. Menuitem "&Volume control options", 40009.

In Windows 10/11, it doesn't exist. But the one to replace is: MENUITEM "&Troubleshoot sound problems",  40008

It's not in "String table" according to Resource Hacker but in "Menu".

**Reply by Anixx** [reply]
Maybe just add an item in Explorer Patcher settings with a link to sndvol -p so to configure this, like there is already a link to the configuration of tray icons?

**Reply by valinet** [reply]
We're not replacing that, as I said, `sndvol -p` doesn't make any sense for people not using the classic volume flyout, so an option to configure the classic volume flyout should show only when using that. I will add a new option in the menu before "Troubleshoot...", but I need a string from somewhere for its name. Oh, and maybe their portal is wrong, maybe that is removed in Windows 10/11. Idk, from somewhere in the current system files I need a proper string.

**Reply by valinet** [reply]
That tool just dumps from the string table. If not in there, it can't do much. If it's in a menu, we can load that as well and extract the string from it, sure.

**Reply by Anixx** [reply]
I suggest to use the caption of that configure dialog itself. In Russian it says "Volume settings": 
![изображение](https://user-images.githubusercontent.com/1271736/138560030-f8a32f2e-eefc-4dc3-9b68-ee1fd79e58a4.png)

**Reply by valinet** [reply]
I will replace that item and make a separate setting for that, it's just easier. Adding new items is a pain in the butt when the menu is skinned.

**Reply by valinet** [reply]
I remembered why I did it like so for the network icon.

**Reply by Anixx** [reply]
For some reason, this option is not yet in the latest build.

Also, I've noticed that the option "Enable missing tray icons" is present in both taskbar and tray tabs.

**Reply by valinet** [reply]
The “reason” is because I did not have time to implement it and also haven’t found a proper way to implement it: adding stuff to the menu is very complicated due to the owner drawing, also changing an item’s name requires research into how the internal structure of the owner drawn menu looks because the name is taken from there; and generally I did not have that insane amount of time to waste on studying yet some more disassembly, again, after 2 months of doing this basically daily. Long story short.

I never said it will be in the next version, but still, maybe I sounded too enthusiastic, I apologize there for giving false hopes, if the case. But still…

And no, I won’t redirect “Troubleshoot…” without changing the label because it is confusing for users. But again, this is relatively easily fixable if you build the software for yourself with just this patch baked in, if it’s really that necessary. Or idk, maybe someone steps in and contributes something in this area.

Also, what you noticed about the “Enable systray icons” link is deliberate. It’s there in 2 places to actually be noticed. I wish more people would do. It’s a “fix” for the billions of questions asking how to enable the “clock, network and sound” icons. It belongs to the “System tray” section, but since that’s not the default page, I moved it to the main page with the hope of it being noticed more easily there. Alas, my hopes are mostly shattered, but nevertheless, maybe it helps 2% of people, I am still happy and relieved.

**Reply by Anixx** [reply]
Thanks for the comment. Is changing the menus in File Explorer more easy? For instance (hypothetically, I do not mean I request this), moving the "Go" submenu from inside "View" to a separate pull-down item as in Win95?

**Reply by valinet** [reply]
I don’t know, haven’t looked on that. Probably easier since the actual data is maybe stored in the MENUITEM structure this time. But who knows for sure how the logic is around that and what one would break with such a patch without actually looking, which personally I certainly haven’t. You could try and see what results you get.

Idk, what I think still misses at this point from the classic theme is a much thinner address bar. This in general is a problem with File Explorer these days on all themes. On high DPI big screens is actually more helpful since it is easier to see, but indeed, on 1366x768 it probably sucks. I’d look on that, but I ain’t got that much time at the moment. I can’t possibly think they changed the height too overly complicated. Isn’t adjusting the height of the Rebar32 control that hosts the address bar enough to make it back thin again, I wonder…?

**Comment by Anixx** [comment]
By the way, there is the first item in the tray context menu "open sound settings" which leads to the settings app. Maybe it would be better to replace this?

**Reply by Gaurav-Original-ClassicShellTester** [reply]
Oh yes that's an idea too! Because there's also right click -> Sounds which leads to the good old mmsys.cpl. Whether to replace "Troubleshoot" or UWP Settings app menu item - either way is fine. Or how Valentin did the trick for the Network icon's context menu, he can redirect "Open Sound settings" also back to mmsys.cpl Playback tab.

Anyway it needs to be in the current OS. Older OS files won't do. I will find if it's there in any of the current files since that Language Portal says so.

**Comment by Gaurav-Original-ClassicShellTester** [comment]
OK I found this using Sysinternals Strings.exe.

Strings C:\Windows\system32\en-US\ *.mui | findstr /i Volume > VolumeOp.txt

It is in Windows 10's sndvol.exe.mui and it is in Resource: DIALOG as a caption for that dialog (which Anixx suggested we use). Earlier one I found was only in Windows 7's sndvolsso.dll.mui as a MenuItem.

In C:\Windows\system32\en-US\sndvol.exe.mui

Dialog 203: 1033 CAPTION "Volume Control Options"

Is that useful?

**Reply by valinet** [reply]
It's what @Anixx suggested we use earlier. You already found it even earlier, Resource Hacker can show those as well. I'll use that, yeah.

**Reply by Anixx** [reply]
For some reason I have no sndvol.exe.mui in en-US folder.

**Reply by valinet** [reply]
@Anixx Do you have English OS language installed? If not, you don't have that file, indeed.

**Comment by plofhaan** [comment]
This old registry tweak still worked for me:

reg add "HKEY_LOCAL_MACHINE\Software\Microsoft\Windows NT\CurrentVersion\MTCUVC" /f /v EnableMtcUvc /t REG_DWORD /d 0
